src/MPDevice.cpp

```diff
@@ -481,6 +481,13 @@ void MPDevice::sendDataDequeue()
         qDebug() << "Full packet#" << i++ << ": " << a;
 #endif
 
+        if (0 == i%5)
+        {
+            /* Added a 10 msec delay for long messages
+             * after every 5th packet to avoid packet drop
+             */
+            QThread::msleep(10);
```

Comment:
After discussing with limpkin about this issue, we need to have a correct fix for this. QThread::msleep is prohibited and a really bad solution.

Actually there is 2 solutions:
- implement throttling when sending packet here in `MPDevice::sendDataDequeue()`. Delay all paquet sending by at least a few ms between each. Problem is it will complexify even more this code (which already does a lot like packet retry, timeout, etc...)

- Implement a buffer in all MPDevice_xxx that waits for the data to be written by the system to the device. All OSs have this API available, and we use non blocking apis right now.

In Linux we do not wait for any data write events. The `QSocketNotifier` is created for `write()` but not used:
https://github.com/mooltipass/moolticute/blob/244c1796779762a42ed3d59ea6c53636c12ee071/src/MPDevice_linux.cpp#L55
Here we need to wait for that signal before writing the next data.

On Windows, this is the same, there is a `QWinOverlappedIoNotifier` but it's only checked for Read, write is being omited:
https://github.com/mooltipass/moolticute/blob/244c1796779762a42ed3d59ea6c53636c12ee071/src/MPDevice_win.cpp#L306

On Macos it's even simpler, because we use a second thread for writing the data as the API is blocking. We can just emit a signal after the write in the thread to notifiy the main thread to continue the write:
https://github.com/mooltipass/moolticute/blob/244c1796779762a42ed3d59ea6c53636c12ee071/src/MPDevice_mac.cpp#L108

Reply:
I implemented a writeBuffer solution in the latest commit for Windows.
I was wonder if race condition can be an issue here: 
https://github.com/mooltipass/moolticute/pull/473/files#diff-e9a41bdf95cd0a995d2c81602f7c5c00R461
https://github.com/mooltipass/moolticute/pull/473/files#diff-e9a41bdf95cd0a995d2c81602f7c5c00R5155
IIRC daemon's MPDevice related code is only running on one thread, so it should not be a problem.

Reply:
I checked your implementation. I'm not a fan, because you added a new queue in MPDevice. It would be better to put the queue in `MPDevice_win` and wait for the write signal there to continue sending data from the buffer like in `MPDevice_linux`:

https://github.com/mooltipass/moolticute/blob/173be4fc19c156cb2a67482dfb74662f2e84e3c9/src/MPDevice_linux.cpp#L102

It's better to have those mechanism in the platform code as it depends on the platform implementation and also `MPDevice` is already complicated enough to maintain another level of buffering there...

Reply:
I can implement the queue in every platform's `MPDevice_xxx`, but I think the queue implementation is a generic one and only platform specific part is sending a request for the next data, when write is finished. (`emit platformWriteFinished()`)

Reply:
> In Linux we do not wait for any data write events. The `QSocketNotifier` is created for `write()` but not used:
> 
> https://github.com/mooltipass/moolticute/blob/244c1796779762a42ed3d59ea6c53636c12ee071/src/MPDevice_linux.cpp#L55
> 
> Here we need to wait for that signal before writing the next data.
> 
I am checking Linux currently.  I connected `sockNotifWrite`'s activated signal for a slot, but the signal is never actually called after writes. I need further investigation regarding this.

Reply:
I have moved the writeBuffer to `MPDevice_win`.
Also tested Linux and Mac there are no issues with writing longer messages. I tested with adding 20-30 credentials in MMM multiple times and have not experienced stuck.